src/sections/upload-dataset-files/UploadDatasetFiles.tsx

```diff
@@ -1,19 +1,104 @@
-import { useEffect } from 'react'
+import { useEffect, useState } from 'react'
 import { FileRepository } from '../../files/domain/repositories/FileRepository'
 import { useLoading } from '../loading/LoadingContext'
 import { useDataset } from '../dataset/DatasetContext'
 import { PageNotFound } from '../page-not-found/PageNotFound'
 import { BreadcrumbsGenerator } from '../shared/hierarchy/BreadcrumbsGenerator'
+import { FileUploader } from './FileUploader'
+import { FileUploadTools } from '../../files/domain/models/FileUploadState'
+import { uploadFile } from '../../files/domain/useCases/uploadFile'
+import { useTranslation } from 'react-i18next'
 
 interface UploadDatasetFilesProps {
   fileRepository: FileRepository
 }
 
-export const UploadDatasetFiles = ({
-  fileRepository: _fileRepository
-}: UploadDatasetFilesProps) => {
+export const UploadDatasetFiles = ({ fileRepository: fileRepository }: UploadDatasetFilesProps) => {
+  const { t } = useTranslation('files')
   const { setIsLoading } = useLoading()
   const { dataset, isLoading } = useDataset()
+  const [fileUploaderState, setState] = useState(FileUploadTools.createNewState([]))
+  const [uploadingToCancelMap, setUploadingToCancelMap] = useState(new Map<string, () => void>())
+  const [uploadFinished, setUploadFinished] = useState(new Set<string>())
+  const [semaphore, setSemaphore] = useState(new Set<string>())
+
+  const timeout = (delay: number) => new Promise((res) => setTimeout(res, delay))
+  const limit = 6
+
+  const acquireSemaphore = async (file: File) => {
```

Comment:
Could you explain me how this semaphore technique works and why we need it? Looks really nice!

Reply:
I hope it clarifies it somewhat, if not we could discuss it in a meeting. I also do not mind deleting it, or removing some other parts of the code, if it is too much, etc.

Reply:
Thank you very much for the explanation. Taking care of CPU performance and controlling processes in such a granular way is the way to go! 
I think it's great, you're looking at the big picture and we're going to need all those features!
👏🏼 👏🏼